### PR TITLE
Collections: PageBean

### PR DESCRIPTION
**Pull request by denisdimitrov**

### CODE REVIEW

**Comment by denisdimitrov** [comment]
Ready.

File: .idea/compiler.xml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (EmilGeorgiev) [inline .idea/compiler.xml]: this should be in .gitignore

File: .idea/copyright/profiles_settings.xml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<component name="CopyrightManager">
```
Comment (EmilGeorgiev) [inline .idea/copyright/profiles_settings.xml]: .

File: .idea/misc.xml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (EmilGeorgiev) [inline .idea/misc.xml]: .

File: .idea/modules.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (EmilGeorgiev) [inline .idea/modules.xml]: .

File: src/pagebean/Demo.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        List<String> list = new LinkedList<>();
+        list.add("First");
+        list.add("Second");
+        list.add("Third");
+        list.add("Fourth");
+        list.add("Fifth");
+        list.add("Sixth");
+        list.add("Seventh");
+        list.add("Eighth");
+        PageBean pb = new PageBean(list,3);
```
Comment (EmilGeorgiev) [inline src/pagebean/Demo.java]: При стартиране на програмата да се показва първата страница от списъка чрез извикване на next().

```go
PageBean pb = new PageBean(list,3);
pg.netx()
```

File: src/pagebean/Demo.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            }
+            if (s.equals(first)){
+                System.out.println(pb.firstPage());
+            }
+            if (s.equals(last)){
+                System.out.println(pb.lastPage());
+            }
+            if (s.equals(curr)){
+                System.out.println(pb.getCurrentPageNumber());
+            }
```
Comment (EmilGeorgiev) [inline src/pagebean/Demo.java]: What will happen if i Scanner read unknown command. I think that client should understand for this as view message.

File: src/pagebean/FirstPageException.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package pagebean;
+
+/**
+ * Created by clouway on 08.07.16.
+ *
+ * @author Denis Dimitrov(@denis.k.dimitrov@gmail.com)
+ */
+public class FirstPageException extends RuntimeException {
```
Comment (EmilGeorgiev) [inline src/pagebean/FirstPageException.java]: Combine ```FirstPageException``` and ```LastPageException``` in ```PageBeenException``` and pass appropriate message in constructor of this class.

```java
public class PageBeenException extends RuntimeException {
    public PageBeenException(String message) {
        super(message)
    }
}
```

Here for example message can be "No more pages, this is last!"

File: src/pagebean/PageBean.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            return list.subList(currentPage * pageSize, currentPage * pageSize + pageSize);
+        }
+    }
+    private boolean hasNext() {
+        if (currentPage * pageSize + pageSize < list.size()){
+            return true;
+        } else {
+            return false;
+        }
+    }
```
Comment (EmilGeorgiev) [inline src/pagebean/PageBean.java]:
```java
private boolean hasNext() {
    return currentPage < pageSize -1
}
```

File: src/pagebean/PageBean.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    public List<T> lastPage() {
+        if (list.size() % pageSize == 0){
+            currentPage = list.size() / pageSize - 1;
+        } else {
+            currentPage = list.size() / pageSize;
+        }
+        return getCurrentPageNumber();
+    }
+
+    public List<T> getCurrentPageNumber() {
```
Comment (EmilGeorgiev) [inline src/pagebean/PageBean.java]: ```getCurrentpageNumber``` is missleading. From this name i understand that if invoke this method i will obtain number     of page, but actually it return all elements in page.

File: src/pagebean/PageBean.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        if (currentPage * pageSize + pageSize > list.size()) {
```
Comment (EmilGeorgiev) [inline src/pagebean/PageBean.java]: here instead this ```if``` you can invoke method ```hasNext()```. We should avoid duplication in our code.

This ```if (currentPage * pageSize + pageSize > list.size()) {``` occurs in two place in your code.

File: src/pagebean/PageBean.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+            return false;
+        }
+    }
+
+    private boolean hasPrevious() {
+        if (currentPage > 0){
+            return true;
+        } else {
+            return false;
+        }
```
Comment (EmilGeorgiev) [inline src/pagebean/PageBean.java]:
```java
private boolean hasPrevious() {
    return currentPage > 0
}
```

File: src/pagebean/Demo.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        System.out.println("Commands: next, prev, first, last, curr");
+        System.out.println("'next' for the next page.");
+        System.out.println("'prev' for the previous page.");
+        System.out.println("'first' for the first page.");
+        System.out.println("'last' for the last page.");
+        System.out.println("'curr' for the current page number.");
+        while (true) {
+            Scanner sc = new Scanner(System.in);
+            String s = sc.nextLine();
+            switch (s.toLowerCase()) {
```
Comment (EmilGeorgiev) [inline src/pagebean/Demo.java]: What will happen if i as client enter command ```nextt``` with typo. I will expect application to tell me that this command does not exist.

File: src/pagebean/PageBean.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    public List<T> lastPage() {
+        if (list.size() % pageSize == 0) {
+            currentPage = list.size() / pageSize - 1;
+        } else {
+            currentPage = list.size() / pageSize;
+        }
+        return getCurrentPageNumber();
+    }
+
+    public List<T> getCurrentPageNumber() {
```
Comment (EmilGeorgiev) [inline src/pagebean/PageBean.java]: getCurrentPageNumber()  should return number of current page not elements in page

File: src/pagebean/PageBean.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        this.list = list;
+        this.pageSize = pageSize;
+    }
+
+    public List<T> next() {
+        if (hasNext()) {
+            currentPage++;
+            return getCurrentPageNumber();
+        }
+        throw new PageBeanException();
```
Comment (EmilGeorgiev) [inline src/pagebean/PageBean.java]: In constructor of this class we should pass message which will be show on user. For example "No more pages."

File: src/pagebean/PageBeanException.java [hunk 15]
```diff
@@ -0,0 +1,13 @@
+package pagebean;
+
+/**
+ * Created by clouway on 08.07.16.
+ *
+ * @author Denis Dimitrov(@denis.k.dimitrov@gmail.com)
+ */
+public class PageBeanException extends RuntimeException {
+    public PageBeanException(){
+        super("No more pages left!");
```
Comment (EmilGeorgiev) [inline src/pagebean/PageBeanException.java]: this message should be pass as parameter of constructor

File: src/pagebean/PageBean.java (hunk 10 above)
Comment (denisdimitrov) [inline src/pagebean/PageBean.java]: Where is the second time? I can't see it.

Comment (EmilGeorgiev) [inline src/pagebean/PageBean.java]: We not have duplication of code. My bad

**Comment by denisdimitrov** [comment]
Ready.

**Comment by EmilGeorgiev** [comment]
LGTM

**Pull request merged**